site3/website/src/pages/bps/BP-66-support-throttling-for-zookeeper-read-of-rereplication.md

```diff
@@ -0,0 +1,26 @@
+# BP-66: support throttling for zookeeper read of rereplication
+
+### Motivation
+
+Each time the cluster trigger the re-replication, all replicators will read data from zookeeper. This can cause a great pressure on Zookeeper. We need to support throttling for zookeeper read of re-replication.
+
+For example, in a Pulsar cluster, we enable auto-recovery for every bookie. There are 400 bookies in a cluster, which means there are 400 replicators in the cluster.
+And there are about 3000 ledgers in each bookie, 1/3 of them are small ledgers, whose size is less than 0.1MB, that is 1000 small ledgers in each bookie.
+If we decommission one bookie, the read latency of zookeeper will increase to minutes. 
+
+
+### Configuration
+add the following configuration:
```

Comment:
Thank you very much for your work. Currently I maintain a bookie cluster with 200 nodes. I applied the following speed limit PR. The bookie process disabled autorecovery and deployed about 10 AutoRecovery processes independently.
https://github.com/apache/bookkeeper/pull/2778

So far, the cluster service operation and maintenance work are relatively good. I think you can separate the AutoRecovery service and set the corresponding replication limit, which may help you.

For the work of this proposal, by individually limiting the frequency of reading zk, the zk service can be reasonably limited and protected, but it is not convenient to limit the byte rate of reading and copying entries because the size of the entry changes;
On the contrary, I feel that PR #2778 can protect zk's read speed through speed limiting.

Let's hear what others have to say.

Reply:
We enable `replicationRateByBytes` to only 3MB as we have 100+ replicators, but the latency of zk read is still very high to minutes level.
![image](https://github.com/apache/bookkeeper/assets/52550727/04372e9a-e39b-4dab-a869-b274578691da)

Limitting the byte rate of replication can't relieve the pressure of zk, but avoid the replication throughput of replication is too high that impact the normal client throughput.

Reply:
If the throughput of a single RW is only 3MB, consider reducing the number of RWs to 10 and adjusting the throughput of a single RW to 30MB. There is no need to maintain 100 replicators. It is generally recommended to separate AR and Bookie.

Reply:
- ZK can't scale out. As the Pulsar & Bookkeeper cluster scale out, the number of replicators must scale out inevitably. When the number of replicators reach to tens or hundreds, the latency will soar to unacceptable level.
- For the convenience of operation and maintenance, we always set `autoRecoveryDaemonEnable=true` for every bookie. We do not adopt the other two options:
  - Deploying another cluster for AutoRecovery increase the complexity of the whole system. 
  - Make small part of bookies in the Bookkeeper cluster work as replicator and set a high value of `replicationRateByBytes` is dangerous. because the throughput of normal client will be impacted by the replication throuhput.
- Actually, there is not a way to relieve the pressure of zk in replicatioin currently.